sdks/python/apache_beam/io/gcp/gcsio.py

```diff
@@ -296,160 +234,87 @@ def delete_batch(self, paths):
     """
     if not paths:
       return []
-
-    paths = iter(paths)
+    if len(paths) > MAX_BATCH_OPERATION_SIZE:
+      raise TooManyRequests("Batch larger than %s", MAX_BATCH_OPERATION_SIZE)
     result_statuses = []
-    while True:
-      paths_chunk = list(islice(paths, MAX_BATCH_OPERATION_SIZE))
-      if not paths_chunk:
-        return result_statuses
-      batch_request = BatchApiRequest(
-          batch_url=GCS_BATCH_ENDPOINT,
-          retryable_codes=retry.SERVER_ERROR_OR_TIMEOUT_CODES,
-          response_encoding='utf-8')
-      for path in paths_chunk:
-        bucket, object_path = parse_gcs_path(path)
-        request = storage.StorageObjectsDeleteRequest(
-            bucket=bucket, object=object_path)
-        batch_request.Add(self.client.objects, 'Delete', request)
-      api_calls = batch_request.Execute(self.client._http)  # pylint: disable=protected-access
-      for i, api_call in enumerate(api_calls):
-        path = paths_chunk[i]
+    with self.client.batch():
+      for path in paths:
+        bucket_name, blob_path = parse_gcs_path(path)
+        bucket = self.client.get_bucket(bucket_name)
+        blob = storage.Blob(blob_path, bucket)
         exception = None
-        if api_call.is_error:
-          exception = api_call.exception
-          # Return success when the file doesn't exist anymore for idempotency.
-          if isinstance(exception, HttpError) and exception.status_code == 404:
+        try:
+          blob.delete()
```

Comment:
This is in a batch environment where persumably request is executed in with's `__exit__`. Have you verified that if there is any NotFound exception, it emits here in the with block or in `__exit__`?

Reply:
I'm fairly sure based on my interpretation of this method: https://github.com/googleapis/python-storage/blob/2b449cd289cb573ca1653bba37ecb84d35a025ad/google/cloud/storage/batch.py#L222, but we don't explicitly test this, so I'll add a test just to make sure

Reply:
The method you pointed is called in finish() which is called in `__exit__`, that is when with block ends. This try - catch block here won't work as intended as it is before batch's `__exit__`

Reply:
Thanks, however I don't think `raise_exception=False` would help much. What I was saying is that in a with clause, the requests are first stacked, and get processed at `__exit__` (see the above link to batch.py). So the following try clause won't catch any http error:

```
with batch()
  try:
    copy_blob
  except err:
    # This won't work
```

because the actual http calls happens in with's `__exit__`.

If we still need to return the copy/delete status of every blob, we then need to process the response in our own. Currently it is done here: https://github.com/googleapis/python-storage/blob/2b449cd289cb573ca1653bba37ecb84d35a025ad/google/cloud/storage/batch.py#L238

Reply:
It's not difficult to re-implement. In the client library it has the all information you need at the code path I pinned (https://github.com/googleapis/python-storage/blob/2b449cd289cb573ca1653bba37ecb84d35a025ad/google/cloud/storage/batch.py#L238) what we would need here isimplement a batch, not using a with clause, get the response and check every copy/delete

Please avoid to introduce regression. When it comes to the need of removing a bunch of unit test it's not going quite right

Reply:
I considered that but all the fields on the client library Batch object are marked as private, so I wasn't sure we could rely on them in the future.

Are the methods in GcsIO being accessed directly by anyone or are they only accessed through GCSFileSystem? If it's the former, I agree that maintaining their return values is important, but if it's the latter I don't feel like that's as high of a priority, especially if we can only do that by using internal variables that aren't guaranteed to remain backwards compatible.

Reply:
thanks, if it is the case that batch copy and batch delete is not actually used, I would suggest make them just call gcsio.copy and gcsio.delete one by one at this moment, to avoid functioning regression. and leave performance regression (use actual batch) as near follow up. At the same time file a Feature Request to gcs client  that make a batch that can return all request reaults.